File: backpack.py
def print_backpack(backpack):
    for a in backpack:
        print("       ".join(list(map(str, a))))
    print()

def theft(product, backpack, S):
    if S < 0:
        return 0
    max_value = 0
    index = -1
    
    for i in range(len(product)):
        if S - product[i][0] >= 0:
            temp = backpack[S - product[i][0]][0] + product[i][1]
            if max_value < temp:
                max_value = temp
                index = i
        elif i == 0:
            return 0

    backpack[S][0] += max_value

    for i in range(1, len(product) + 1):
        backpack[S][i] = backpack[S - product[index][0]][i]

    backpack[S][index + 1] += 1
    print_backpack(backpack)
    return backpack[S][0]


def main():
    S = 20
    product = [
        [3, 5],
        [4, 7],
        [5, 10]
    ]
    backpack = [[0] * 4 for _ in range(S + 1)]

    for i in range(S + 1):
        theft(product, backpack, i)

    for j in range(4):
        print(backpack[S][j], end=" ")
    print()


if __name__ == "__main__":
    main()

File: backpack.cpp
#include <iostream>
#include <vector>

using namespace std;

typedef vector<vector<int>> Vector;

int theft(Vector& product, Vector& backpack, int S) {
	if (S < 0)
		return 0;
	int max = 0, index{};
	for (int i = 0; i < product.size(); i++)
		if (S - product[i][0] > -1) {
			int temp = backpack[S - product[i][0]][0] + product[i][1];
			if (max < temp) {
				max = temp;
				index = i;
			}
		}
		else if (i == 0)
			return 0;
	backpack[S][0] += max;
	for (int i = 1; i < product.size() + 1; i++)
		backpack[S][i] = backpack[S - product[index][0]][i];
	backpack[S][index + 1]++;
	return backpack[S][0];
}

int main() {
	int S = 23;
	Vector product = {
		{5,9},
		{7,13},
		{11,21}
	};
	Vector backpack(S + 1, vector<int>(4));

	for (int i = 0; i < S + 1; i++)
		theft(product, backpack, i);

	//for (int i = 0; i < S + 1; i++) {
		//cout << i << ": ";
		for (int j = 0; j < 4; j++)
			cout << backpack[S][j] << " ";
		cout << endl;
	//}
}